xls/modules/zstd/zstd_dec_test.cc

```diff
@@ -0,0 +1,241 @@
+// Copyright 2020 The XLS Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include <fstream>
+#include <memory>
+
+#include "gtest/gtest.h"
+#include "xls/common/file/filesystem.h"
+#include "xls/common/file/get_runfile_path.h"
+#include "xls/common/status/matchers.h"
+#include "xls/interpreter/interpreter_proc_runtime.h"
+#include "xls/interpreter/serial_proc_runtime.h"
+#include "xls/ir/events.h"
+#include "xls/ir/ir_parser.h"
+#include "xls/modules/zstd/data_generator.h"
+#include "zstd.h"
+
+namespace xls {
+namespace {
+
+class ZstdDecodedPacket {
+ public:
+  static std::optional<ZstdDecodedPacket> MakeZstdDecodedPacket(Value packet) {
+    // Expect tuple
+    if (!packet.IsTuple()) return std::nullopt;
+    // Expect exactly 3 fields
+    if (packet.size() != 3) return std::nullopt;
+    for (int i = 0; i < 3; i++) {
+      // Expect fields to be Bits
+      if (!packet.element(i).IsBits()) return std::nullopt;
+      // All fields must fit in 64bits
+      if (!packet.element(i).bits().FitsInUint64()) return std::nullopt;
+    }
+
+    std::vector<uint8_t> data = packet.element(0).bits().ToBytes();
+    absl::StatusOr<uint64_t> len = packet.element(1).bits().ToUint64();
+    if (!len.ok()) return std::nullopt;
+    uint64_t length = *len;
+    bool last = packet.element(2).bits().IsOne();
+
+    return ZstdDecodedPacket(data, length, last);
+  }
+
+  std::vector<uint8_t> GetData() { return data; }
+
+  uint64_t GetLength() { return length; }
+
+  bool IsLast() { return last; }
+
+  const std::string PrintData() const {
+    std::stringstream s;
+    for (int j = 0; j < sizeof(uint64_t) && j < data.size(); j++) {
+      s << "0x" << std::setw(2) << std::setfill('0') << std::right << std::hex
+        << (unsigned int)data[j] << std::dec << ", ";
+    }
+    return s.str();
+  }
+
+  friend std::ostream& operator<<(std::ostream& os,
+                                  const ZstdDecodedPacket& packet) {
+    return os << "ZstdDecodedPacket { data: {" << packet.PrintData()
+              << "}, length: " << packet.length << " last: " << packet.last
+              << "}" << std::endl;
+  }
+
+ private:
+  ZstdDecodedPacket(std::vector<uint8_t> data, uint64_t length, bool last)
+      : data(data), length(length), last(last) {}
+
+  std::vector<uint8_t> data;
+  uint64_t length;
+  bool last;
+};
+
+class ZstdDecoderTest : public ::testing::Test {
```

Comment:
would be useful to test the JIT'ed version as well, might be easier to do once we have proc based FFI #1301.